File: Algorithms/Merge Sort/Merge sort.cpp
	 #include<bits/stdc++.h>
	 using namespace std;
    vector<int> merge(vector<int> left,vector<int> right)
    {
        int l=0,r=0;
        vector<int> ans;
        while(l<left.size() && r<right.size())
        {
            if(left[l]<right[r])
            {ans.push_back(left[l]);
            l++;}
            else
            if(left[l]>right[r])
                {ans.push_back(right[r]);
                r++;}
            else
            if(left[l]==right[r])
                {ans.push_back(left[l]);
                ans.push_back(left[l]);
                r++;l++;}
        }
                   
        
                   while(l<left.size())
                   {
                       ans.push_back(left[l]);
                       l++;
                   }
                   while(r<right.size())
                   {
                       ans.push_back(right[r]);
                       r++;
                   }
                   return ans;
    }
    vector<int> mergesort(vector<int> nums)
    {
        if(nums.size()==1)
            return nums;
        
        int m=nums.size()/2;
        
        vector<int> left;
        for(int i=0;i<m;i++)
            left.push_back(nums[i]);
        
        left=mergesort(left);
        
        vector<int> right;
        for(int i=m;i<nums.size();i++)
            right.push_back(nums[i]);
        
        right=mergesort(right);
        
        return merge(left,right);
    }
 
    
    int main()
    {
     int n;
     cin>>n;
     vector<int> nums(n);
     for (int i = 0; i < n; ++i)
     cin>>nums[i];
    
      vector<int> res=mergesort(nums); 
      for (int i = 0; i < n; ++i)
      cout<<res[i]<<" ";

      return 0; 
    }

File: Algorithms/Merge Sort/MergeSort.cs
using System;
namespace MergeSort
{
    class Program
    {
        static public void MergeMethod(int[] numbers, int left, int mid, int right)
        {
            int[] temp = new int[25];
            int i, left_end, num_elements, tmp_pos;
            left_end = (mid - 1);
            tmp_pos = left;
            num_elements = (right - left + 1);
            while ((left <= left_end) && (mid <= right))
            {
                if (numbers[left] <= numbers[mid])
                    temp[tmp_pos++] = numbers[left++];
                else
                    temp[tmp_pos++] = numbers[mid++];
            }
            while (left <= left_end)
                temp[tmp_pos++] = numbers[left++];
            while (mid <= right)
                temp[tmp_pos++] = numbers[mid++];
            for (i = 0; i < num_elements; i++)
            {
                numbers[right] = temp[right];
                right--;
            }
        }
        static public void SortMethod(int[] numbers, int left, int right)
        {
            int mid;
            if (right > left)
            {
                mid = (right + left) / 2;
                SortMethod(numbers, left, mid);
                SortMethod(numbers, (mid + 1), right);
                MergeMethod(numbers, left, (mid + 1), right);
            }
        }
        static void Main(string[] args)
        {
            int[] numbers = { 38, 27, 43, 3, 9, 82, 10 };
            int len = numbers.Length;
            Console.WriteLine("Before Merge Sort:");
            foreach(int item in numbers)
            {
                Console.Write(item + " ");
            }
            Console.WriteLine();
            Console.WriteLine("After Merge Sort");
            SortMethod(numbers, 0, len - 1);
            foreach (int item in numbers)
            {
                Console.Write(item + " ");
            }
            Console.Read();
        }
    }
}
